compiler/luci/pass/src/FuseMulWithConvPass.cpp

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2023 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/FuseMulWithConvPass.h"
+
+#include "helpers/NodeFiller.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+#include <luci/Service/CircleNodeClone.h>
+#include <luci/Service/Nodes/CircleConst.h>
+
+namespace
+{
+
+#define RETURN_FALSE_UNLESS(cond) \
+  if (not(cond))                  \
+    return false;
+
+inline uint32_t cal_offset(const luci::CircleConst *node, const uint32_t *indices)
+{
+  return indices[0] * node->dim(1).value() * node->dim(2).value() * node->dim(3).value() +
+         indices[1] * node->dim(2).value() * node->dim(3).value() +
+         indices[2] * node->dim(3).value() + indices[3];
+}
+
+/**
+ *  Fuse Mul with Conv if possible
+ *
+ *  NOTE: In case mul is channewise constant, we can try to merge mul with nconv,
+ *
+ *  BEFORE
+ *                    |
+ *              [CircleConv2D] (no activation)
+ *                    |
+ *                  [Mul] (channel-wise constant)
+ *                    |
+ *
+ *  AFTER
+ *                    |
+ *            [CircleConv2D] (with updated kernels, bias, and activation)
+ *                    |
+ *
+ */
+
+bool fuse_mul_with_conv(luci::CircleMul *mul)
+{
+  // sanity check
+  RETURN_FALSE_UNLESS(mul->dtype() == loco::DataType::FLOAT32);
+
+  luci::CircleConst *const_mul_operand = nullptr;
+  luci::CircleConv2D *conv = nullptr;
+  RETURN_FALSE_UNLESS(luci::fill(&const_mul_operand, &conv).with_commutative_args_of(mul));
+
+  // sanity check
+  RETURN_FALSE_UNLESS(conv->dtype() == loco::DataType::FLOAT32 &&
+                      const_mul_operand->dtype() == loco::DataType::FLOAT32);
+
+  //  NOTE for general activation function: S * Act(A * B) != Act(A*(SB))
+  if (conv->fusedActivationFunction() != luci::FusedActFunc::NONE)
+    return false;
+
+  // check that const_mul_operand is channel-wise constant
+  RETURN_FALSE_UNLESS(const_mul_operand->rank() == 4);
+
+  // check channel dimension will be checked later
+  RETURN_FALSE_UNLESS(const_mul_operand->dim(0).value() == 1 &&
+                      const_mul_operand->dim(1).value() == 1 &&
+                      const_mul_operand->dim(2).value() == 1);
```

Comment:
I think this condition can be easily extended.

Can you make this work for "rank-1 scalar" and "rank-4" channel-wise const?

Reply:
yes. Ok.

Reply:
done. Added cases for `rank == 1 || rank == 0`.